File: DSA_ProblemSolving_patterns/src/problems/UnionFind/unionFind.ts
/**
 * ============================================================================
 * UNION-FIND / DISJOINT SET UNION (DSU)
 * ============================================================================
 *
 * WHAT IT IS:
 * A structure that tracks a collection of elements partitioned into disjoint
 * groups, supporting two operations blazingly fast:
 *   - find(x):  which group is x in? (returns the group's representative "root")
 *   - union(a, b): merge the groups containing a and b.
 *
 * PATTERN:
 * - **Represent each group as a tree; the root identifies the group.** Two
 *   elements are connected iff they share a root. Merging = point one root at
 *   the other. Two optimizations turn near-linear trees into near-flat ones.
 *
 * WHEN TO USE:
 * - "How many groups / are these connected / detect a cycle in an UNDIRECTED
 *   graph / progressively merge sets." Especially when edges ARRIVE OVER TIME
 *   (online) — DFS/BFS want the whole graph up front; union-find merges
 *   incrementally.
 *
 * REAL-WORLD ANALOGIES:
 * - Network connectivity: are machines A and B on the same network as cables
 *   get plugged in one by one?
 * - Account/friend merging: "same person" clusters as duplicate links appear.
 * - Kruskal's MST: add an edge iff it joins two different components.
 *
 * COMPLEXITY LADDER (n elements, m operations):
 *   Step 1  Naive (root chains can grow linear)     find/union O(n)
 *   Step 2  + Path compression                       ~O(log n) amortized
 *   Step 3  + Union by rank/size ★                   ~O(α(n)) ≈ O(1) amortized
 *           (α = inverse Ackermann — ≤ 4 for any practical n)
 * ============================================================================
 */

/**
 * ----------------------------------------------------------------------------
 * STEP 1 — NAIVE UNION-FIND (the mental model, before optimizing)
 * ----------------------------------------------------------------------------
 * `parent[i]` points at i's parent; a root points at itself. `find` walks up
 * to the root; `union` links one root under the other.
 *
 * Why it's slow: repeated unions can build a long chain (a linked list), so
 * `find` degrades to O(n) — walking the whole chain each time. Steps 2-3 keep
 * the trees shallow.
 */
export class NaiveUnionFind {
  protected parent: number[];

  constructor(size: number) {
    // Everyone starts in their own group: parent[i] = i (i is its own root).
    this.parent = Array.from({ length: size }, (_, i) => i);
  }

  find(x: number): number {
    while (this.parent[x] !== x) x = this.parent[x]; // walk up to the root
    return x;
  }

  union(a: number, b: number): void {
    const rootA = this.find(a);
    const rootB = this.find(b);
    if (rootA !== rootB) this.parent[rootA] = rootB; // hang one root under the other
  }

  connected(a: number, b: number): boolean {
    return this.find(a) === this.find(b);
  }
}

/**
 * ----------------------------------------------------------------------------
 * STEP 2 & 3 ★ — OPTIMIZED UNION-FIND (path compression + union by rank)
 * ----------------------------------------------------------------------------
 * PATH COMPRESSION: while finding the root, re-point every node along the way
 * DIRECTLY at the root — so the next find is nearly O(1). "Flatten as you go."
 *
 * UNION BY RANK: always attach the SHORTER tree under the taller one, so the
 * height grows as slowly as possible (never make the tall tree taller for no
 * reason).
 *
 * Together these give ~O(α(n)) amortized per op — effectively constant.
 *
 * `count` tracks the number of disjoint groups: starts at `size`, drops by one
 * on every union that actually merges two different groups.
 */
export class UnionFind {
  private parent: number[];
  private rank: number[]; // upper bound on tree height rooted at i
  private groupCount: number;

  constructor(size: number) {
    this.parent = Array.from({ length: size }, (_, i) => i);
    this.rank = new Array<number>(size).fill(0);
    this.groupCount = size; // every element is its own group initially
  }

  /** find with PATH COMPRESSION (iterative two-pass: find root, then flatten). */
  find(x: number): number {
    let root = x;
    while (this.parent[root] !== root) root = this.parent[root];

    // Second pass: point everything on the path straight at the root.
    while (this.parent[x] !== root) {
      const next = this.parent[x];
      this.parent[x] = root;
      x = next;
    }
    return root;
  }

  /** union by RANK. Returns true if a merge happened (they were separate). */
  union(a: number, b: number): boolean {
    const rootA = this.find(a);
    const rootB = this.find(b);
    if (rootA === rootB) return false; // already in the same group

    // Attach the shorter tree under the taller one.
    if (this.rank[rootA] < this.rank[rootB]) {
      this.parent[rootA] = rootB;
    } else if (this.rank[rootA] > this.rank[rootB]) {
      this.parent[rootB] = rootA;
    } else {
      this.parent[rootB] = rootA; // equal ranks: pick one, its rank grows by 1
      this.rank[rootA]++;
    }

    this.groupCount--; // two groups just became one
    return true;
  }

  connected(a: number, b: number): boolean {
    return this.find(a) === this.find(b);
  }

  /** Number of disjoint groups remaining. */
  count(): number {
    return this.groupCount;
  }
}

/**
 * ----------------------------------------------------------------------------
 * APPLICATION — Number of Connected Components (LeetCode 323), the DSU way
 * ----------------------------------------------------------------------------
 * Union every edge; the leftover group count is the answer. Compare with the
 * DFS version in Graph/graphPatterns.ts — DSU shines when edges stream in or
 * when you also need "are these two connected?" queries interleaved.
 *
 * DRY-RUN on n=5, edges=[[0,1],[1,2],[3,4]]:
 *   start count=5
 *   union(0,1) → 4 · union(1,2) → 3 · union(3,4) → 2
 *   → 2 components ({0,1,2}, {3,4}) ✓
 *
 * Time: O(n + m·α(n)) ≈ O(n + m). Space: O(n).
 */
export function countComponentsUnionFind(n: number, edges: number[][]): number {
  const uf = new UnionFind(n);
  for (const [a, b] of edges) uf.union(a, b);
  return uf.count();
}

/**
 * APPLICATION — cycle detection in an UNDIRECTED graph.
 * If an edge connects two nodes already in the same group, that edge closes a
 * cycle. (This is the reject/accept test at the heart of Kruskal's MST.)
 *
 * @example
 * hasCycleUndirected(3, [[0,1],[1,2],[2,0]]); // true  (triangle)
 * hasCycleUndirected(3, [[0,1],[1,2]]);       // false (a path)
 */
export function hasCycleUndirected(n: number, edges: number[][]): boolean {
  const uf = new UnionFind(n);
  for (const [a, b] of edges) {
    if (!uf.union(a, b)) return true; // both endpoints already connected → cycle
  }
  return false;
}

/**
 * ============================================================================
 * INTERVIEW NOTES (quick revision)
 * ============================================================================
 * 1. Recognition cue: "connected components / connectivity queries / detect
 *    cycle in undirected graph / merge sets as edges arrive." If edges stream
 *    in over time, prefer DSU over DFS/BFS (which need the whole graph first).
 * 2. The two optimizations are the whole interview: PATH COMPRESSION (flatten
 *    on find) + UNION BY RANK/SIZE (shorter under taller). Quote the result:
 *    ~O(α(n)) amortized, effectively constant.
 * 3. `union` returning "did a merge happen?" is what powers component-counting
 *    and cycle detection — don't discard that boolean.
 * 4. Directed-graph cycle detection is DIFFERENT — use DFS colors or Kahn's
 *    topological sort (Graph/graphPatterns.ts), NOT union-find.
 * 5. Follow-ups: Kruskal's MST, Accounts Merge (LC721), Redundant Connection
 *    (LC684 — the first edge that fails union IS the redundant one), Number of
 *    Islands via DSU.
 */

File: Go/unionfind/unionfind.go
// Package unionfind merges sets and answers connectivity queries.
//
// PATTERN         Union-Find / Disjoint Set Union (DSU)
// WHEN TO USE     "connected components", "is this a valid tree", "cycle in an
// UNDIRECTED graph", "which edge is redundant", and anything where edges arrive
// ONE AT A TIME and you must answer connectivity as you go.
// WASTE IT KILLS  Re-running DFS/BFS after every new edge.
//
// DSU OR DFS? THE DECIDING QUESTION: is the graph fixed or growing?
//   - Fixed graph, count the components once  -> DFS is simpler, just use it.
//   - Edges streaming in, or "which edge closes a cycle" -> DSU. It answers
//     each query in near-constant time instead of re-traversing.
//
// TWO OPTIMISATIONS, AND WHY EACH EXISTS:
//  1. PATH COMPRESSION — after a Find, point every node on the path straight at
//     the root, so the next query is O(1).
//  2. UNION BY RANK — always hang the shorter tree under the taller, so the
//     structure never degenerates into a linked list.
//
// Together they give O(α(n)) amortised — inverse Ackermann, under 5 for any n
// you will ever see. Say "effectively constant"; claiming exactly O(1) is wrong.
package unionfind

// UnionFind tracks disjoint sets over 0..n-1.
type UnionFind struct {
	parent []int
	rank   []int // upper bound on tree height, used to keep trees flat
	count  int   // how many disjoint sets remain
}

// NewUnionFind starts with every element in its own set.
func NewUnionFind(n int) *UnionFind {
	u := &UnionFind{
		parent: make([]int, n),
		rank:   make([]int, n),
		count:  n,
	}
	for i := range n {
		u.parent[i] = i // each element is its own root
	}
	return u
}

// Find returns the representative of x's set, compressing the path on the way.
//
// The loop rewires each node to its GRANDPARENT (path halving) — one pass, no
// recursion, and it flattens the tree just as effectively as the two-pass
// version. This is the whole reason DSU is fast.
func (u *UnionFind) Find(x int) int {
	for u.parent[x] != x {
		u.parent[x] = u.parent[u.parent[x]] // halve the path
		x = u.parent[x]
	}
	return x
}

// Union merges the sets containing a and b.
//
// RETURNS FALSE when they were ALREADY connected — and that return value is the
// entire cycle-detection mechanism: an edge joining two already-connected nodes
// closes a cycle.
func (u *UnionFind) Union(a, b int) bool {
	rootA, rootB := u.Find(a), u.Find(b)
	if rootA == rootB {
		return false // already together: this edge is redundant
	}
	// Hang the shorter tree under the taller one.
	if u.rank[rootA] < u.rank[rootB] {
		rootA, rootB = rootB, rootA
	}
	u.parent[rootB] = rootA
	if u.rank[rootA] == u.rank[rootB] {
		u.rank[rootA]++ // equal heights: the merged tree is one taller
	}
	u.count--
	return true
}

// Connected reports whether a and b are in the same set.
func (u *UnionFind) Connected(a, b int) bool { return u.Find(a) == u.Find(b) }

// Count returns the number of disjoint sets.
func (u *UnionFind) Count() int { return u.count }

// ---------------------------------------------------------------------------
// The three questions this pattern owns
// ---------------------------------------------------------------------------

// CountComponents counts connected components of an undirected graph (LC323).
// Start at n components and subtract one for every union that actually merged.
func CountComponents(n int, edges [][]int) int {
	u := NewUnionFind(n)
	for _, e := range edges {
		u.Union(e[0], e[1])
	}
	return u.Count()
}

// ValidTree reports whether the edges form a tree (LC261).
//
// A TREE IS TWO CONDITIONS AT ONCE, and both are needed:
//  1. no cycle        — every Union must actually merge something
//  2. fully connected — exactly one component at the end
//
// The edge-count shortcut (a tree has exactly n-1 edges) plus "no cycle" is
// equivalent, and worth mentioning as the O(1) pre-check.
func ValidTree(n int, edges [][]int) bool {
	if len(edges) != n-1 {
		return false // too few cannot connect; too many must cycle
	}
	u := NewUnionFind(n)
	for _, e := range edges {
		if !u.Union(e[0], e[1]) {
			return false // this edge closed a cycle
		}
	}
	return u.Count() == 1
}

// FindRedundantConnection returns the edge that, when removed, leaves a tree
// (LC684). Nodes are 1-indexed.
//
// WHY DSU IS THE PERFECT FIT: process edges in order and the FIRST one joining
// two already-connected nodes is the answer. The problem explicitly asks for the
// last such edge in the input, and since exactly one extra edge exists, the
// first cycle-closing edge encountered IS that edge.
func FindRedundantConnection(edges [][]int) []int {
	u := NewUnionFind(len(edges) + 1) // 1-indexed: allocate one spare slot
	for _, e := range edges {
		if !u.Union(e[0], e[1]) {
			return e
		}
	}
	return nil
}
